app/src/main/java/org/thoughtcrime/securesms/profiles/manage/ManageProfileViewModel.java

```diff
@@ -0,0 +1,248 @@
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.badges.models.Badge;
+import org.thoughtcrime.securesms.dependencies.ApplicationDependencies;
+import org.thoughtcrime.securesms.jobs.RetrieveProfileJob;
+import org.thoughtcrime.securesms.mediasend.Media;
+import org.thoughtcrime.securesms.profiles.AvatarHelper;
+import org.thoughtcrime.securesms.profiles.ProfileName;
+import org.thoughtcrime.securesms.providers.BlobProvider;
+import org.thoughtcrime.securesms.recipients.Recipient;
+import org.thoughtcrime.securesms.recipients.RecipientForeverObserver;
+import org.thoughtcrime.securesms.util.DefaultValueLiveData;
+import org.thoughtcrime.securesms.util.FeatureFlags;
+import org.thoughtcrime.securesms.util.SingleLiveEvent;
+import org.thoughtcrime.securesms.util.livedata.LiveDataUtil;
+import org.whispersystems.libsignal.util.guava.Optional;
+import org.whispersystems.signalservice.api.util.StreamDetails;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+class ManageProfileViewModel extends ViewModel {
+
+  private static final String TAG = Log.tag(ManageProfileViewModel.class);
+
+  private final MutableLiveData<InternalAvatarState> internalAvatarState;
+  private final MutableLiveData<ProfileName>         profileName;
+  private final MutableLiveData<String>              username;
+  private final MutableLiveData<String>              about;
+  private final MutableLiveData<String>              aboutEmoji;
+  private final LiveData<AvatarState>                avatarState;
+  private final SingleLiveEvent<Event>               events;
+  private final RecipientForeverObserver             observer;
+  private final ManageProfileRepository              repository;
+  private final MutableLiveData<Optional<Badge>>     badge;
+
+  private byte[] previousAvatar;
+
+  public ManageProfileViewModel() {
+    this.internalAvatarState = new MutableLiveData<>();
+    this.profileName         = new MutableLiveData<>();
+    this.username            = new MutableLiveData<>();
+    this.about               = new MutableLiveData<>();
+    this.aboutEmoji          = new MutableLiveData<>();
+    this.events              = new SingleLiveEvent<>();
+    this.repository          = new ManageProfileRepository();
+    this.badge               = new DefaultValueLiveData<>(Optional.absent());
+    this.observer            = this::onRecipientChanged;
+    this.avatarState         = LiveDataUtil.combineLatest(Recipient.self().live().getLiveData(), internalAvatarState, (self, state) -> new AvatarState(state, self));
+
+    SignalExecutors.BOUNDED.execute(() -> {
+      onRecipientChanged(Recipient.self().fresh());
+
+      StreamDetails details = AvatarHelper.getSelfProfileAvatarStream(ApplicationDependencies.getApplication());
+      if (details != null) {
+        try {
+          internalAvatarState.postValue(InternalAvatarState.loaded(StreamUtil.readFully(details.getStream())));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *getStream* method in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when StreamUtil.readFully throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).